### PR TITLE
Add support for `arrayUnion()` function

### PR DESCRIPTION
**Pull request by petern48**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for `arrayUnion` function

Closes #61553 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit e4ba3b47fc97c503ca3e5168f405ab8e34732baa with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68989/e4ba3b47fc97c503ca3e5168f405ab8e34732baa/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/68989/e4ba3b47fc97c503ca3e5168f405ab8e34732baa/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/03224_arrayUnion.sql [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+select arraySort(arrayUnion([], arr)) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([1,2], arr)) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([1,2], [1,2,3,4])) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([], [])) from array_union order by arr;
+select '-------';
+
+optimize table array_union;
```
Comment (vdimir) [inline tests/queries/0_stateless/03224_arrayUnion.sql:24]: What's `optimize` query for?

Comment (petern48) [reply tests/queries/0_stateless/03224_arrayUnion.sql:24]: I don't know to be honest, probably should've deleted it. I copied over the arrayIntersect test to use as a template and it was there. I'll delete it.

**Comment by vdimir** [comment]
> AST fuzzer (asan) — Logical error: 'Sizes of nested column and null map of Nullable column are not equal'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/68989/693abc941ee6326729d20730fafa89abcfef4c83/ast_fuzzer__asan_.html)
> AST fuzzer (debug) — Assertion `(n >= (static_cast<ssize_t>(pad_left_) ? -1 : 0)) && (n <= static_cast<ssize_t>(this->size()))' failed [Details](https://s3.amazonaws.com/clickhouse-test-reports/68989/693abc941ee6326729d20730fafa89abcfef4c83/ast_fuzzer__debug_.html)
> AST fuzzer (msan) — Logical error: 'Wrong column in Decimal comparison'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/68989/693abc941ee6326729d20730fafa89abcfef4c83/ast_fuzzer__msan_.html)
> AST fuzzer (tsan) — Received signal 11 [Details](https://s3.amazonaws.com/clickhouse-test-reports/68989/693abc941ee6326729d20730fafa89abcfef4c83/ast_fuzzer__tsan_.html)
> AST fuzzer (ubsan) — Received signal 11 [Details](https://s3.amazonaws.com/clickhouse-test-reports/68989/693abc941ee6326729d20730fafa89abcfef4c83/ast_fuzzer__ubsan_.html)


These failures looks related to the change. Could you please check `fatal.log` and look for queries in `<Fatal>` log entries, e.g.


```
2024.08.30 11:11:25.637176 [ 1414 ] {} <Fatal> BaseDaemon: (version 24.9.1.1553, build id: 491DCBCA4BD8C028D9BA4E3660E66F322C88E4E3, git hash: 09ac8a4889f4b89d03005a61cb3f56006faaf7fe) (from thread 174) (query_id: 0a75ec11-cef7-4e42-8577-1704d5238232) (query: SELECT arrayUnion([toNullable(materialize('-------'))]), arraySort(arrayUnion([1, 2])) FROM array_union__fuzz_0 ORDER BY arr ASC) Received signal Aborted (6)
```

**Comment by petern48** [comment]
Is there a way for me to find what content was in the tables generated by the fuzzer? It said the below query failed. I couldn't reproduce the behavior with the array_union table in my test case, so the contents of the array_union__fuzz_3 table in the query must be different.

`SELECT arraySort(arrayUnion(arr)) FROM array_union__fuzz_3 PREWHERE 11 ORDER BY arr DESC`

**Comment by vdimir** [comment]
> Is there a way for me to find what content was in the tables generated by the fuzzer?

You may find table schema in `CREATE` statement in server logs, however, not the data, since it can be large amount and we do not persist it in test run. Usually, generating some corner-case value for specific type can help (e.g. NULL for nullable types, maximal values for int, etc)

File: tests/queries/0_stateless/03224_arrayUnion.sql [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+select '-------';
+select arraySort(arrayUnion(arr, [])) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([], arr)) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([1,2], arr)) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([1,2], [1,2,3,4])) from array_union order by arr;
+select '-------';
+select arraySort(arrayUnion([], [])) from array_union order by arr;
```
Comment (superdiaodiao) [inline tests/queries/0_stateless/03224_arrayUnion.sql:33]: Seems that these lines are the same with line 11 to line 21.

Comment (petern48) [reply tests/queries/0_stateless/03224_arrayUnion.sql:21]: Yep thanks for catching that

**Comment by petern48** [comment]
@vdimir I can't seem to reproduce failing either of these test cases. Do you have any advice? Is it possible that these failures are unrelated to my changes? The failed stateful test (`ProfileEvent_DiskS3ReadRequestsErr`) relates to S3 reads (which seems unrelated) and the failed stateless test  (`02532_send_logs_level_test`) passes fine for me locally.

**Review by vdimir — APPROVED** [review]

**Pull request merged**